### ISSUE TITLE
Structs as input to functions

### ISSUE THREAD

**Issue by calebwin** [calebwin/emu#7]
The only things Emu functions can really accept right now are vectors (technically arrays/pointers) and primitive data (f32 or i32). Simple structures could be accepted with 2 changes.

-A change to the language so you can declare what kind of structs you accept and how to unpackage primitive data from them in the declaration of the Emu function
-A change to the build! macro to generate a function that can accept structs of a certain type and unpackage them into primitive data to send to the Emu function.

Before these changes are implemented, we should think about how the general interface to an Emu user should change. How should they pass structs to functions in a way that is most seamless.

**Comment by calebwin** [comment]
Also, I think vectors of structs should also be allowed as input. Would that make sense?

Note that right now a function in Emu is basically something that is called for each element of a vector at least if you are using the `build!` macro to generate Rust functions that do that for you.

**Comment by kpp** [comment]
> Also, I think vectors of structs should also be allowed as input. Would that make sense?

Yes. Take a look at https://github.com/kpp/test_ocl/blob/master/src/main.rs#L11

**Comment by calebwin** [comment]
Interesting.
```rust
struct Cluster {
    input_ones: ocl::prm::Ulong2,
    input_zeros: ocl::prm::Ulong2,
}
```
I don't think this is how a user should have to define structs. They should just be able to say-

```rust
struct Cluster {
    input_ones: u64,
    input_zeros: u64
}
```

**Comment by kpp** [comment]
`Ulong2` - is a vector data type of two 64-bit unsigned integers. See https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/vectorDataTypes.html.

So Ulong2 != u64, it's like u128.

**Comment by calebwin** [comment]
Ok. By the way, the "vectors" that Emu functions accept are not OpenCL vectors, they are arrays/pointers. So there isn't yet support for vector data types.

**Comment by omac777** [comment]
Strangely enough, when I googled gpu integer, one of the first relevant results was quasar which supports integer types.

https://gepura.io/quasar/quick-reference-manual/
2.2.2  Integer types 

I did notice you mentioned not only supporting f32 but i32 integers also.  I'll give it a try.
I wish there was more support for dealing with string handling within the gpu with emu.  sorting/slicing/regex'ing.
